feat: add fallback dynamic cmd by default

kylo252 opened the pull request:
Make use of `dynamic_command` as a fallback when the base command isn't found. This is particularly useful for node_modules, for example: `prettier`.

```lua
generator_opts = {
    command = "prettier",
    args = h.range_formatting_args_factory({ "--stdin-filepath", "$FILENAME" }),
    to_stdin = true,
    dynamic_command = require("null-ls.helpers.command_resolver").from_node_modules
}
```

Related discussion: https://github.com/jose-elias-alvarez/null-ls.nvim/pull/349#pullrequestreview-807707808

kylo252 commented:
@jose-elias-alvarez, what do you think of this approach? I just prefer this for readability over `prefer_local` for now, but we can also use that instead.

The modularization is more for cleanup than anything else.

jose-elias-alvarez commented:
I like this idea and don't have any issue with the changes. Before making this into a default, I'd like to profile how much overhead this adds, both on the first search and with subsequent function calls. If it's opt-in, I think some amount of overhead is acceptable, but if not, we should think about this a bit more (especially since some of these Node sources are already pretty slow).

If the difference is minimal, we can go ahead and add the default to more Node sources, too.

kylo252 commented:
> I like this idea and don't have any issue with the changes. Before making this into a default, I'd like to profile how much overhead this adds, both on the first search and with subsequent function calls. 

Notice how I kept the caching mechanism, however, that is still done per buffer which is mostly fine but could probably be further improved. Maybe we should do the caching per command or per root, or a combination of all three?
 
> If it's opt-in, I think some amount of overhead is acceptable, but if not, we should think about this a bit more (especially since some of these Node sources are already pretty slow). 

Agreed, and this will probably be least noticeable for formatters anyway, especially if it's done during `on_attach`.

jose-elias-alvarez commented:
A user came up with a solution to a similar problem [here](https://github.com/jose-elias-alvarez/null-ls.nvim/issues/509#issuecomment-1002950065). Ideally, buffers under the same `resolved_cwd` would automatically resolve to the same command, though we'd have to balance that against adding complexity.

kylo252 commented:
> A user came up with a solution to a similar problem [here](https://github.com/jose-elias-alvarez/null-ls.nvim/issues/509#issuecomment-1002950065). Ideally, buffers under the same `resolved_cwd` would automatically resolve to the same command, though we'd have to balance that against adding complexity.

I think my approach should be simpler. I've also added a `cwd` attribute to make things a bit easier, even though it's probably not needed.

I would actually recommend renaming `cwd` to `root_pattern` to make it more consistent across the board, but that would be a breaking change, so I'll leave it to your discretion 😃

jose-elias-alvarez commented:
I'm not necessarily opposed to making a breaking change, since it's a more advanced option, but I'd be worried about confusion with lspconfig's `root_pattern`, which is different. I think `cwd` also makes it clearer that it's a per-source option.

I briefly looked at the most recent `per_workspace` changes, and I think I'd prefer to keep the original logic for now and improve on it later if we run into performance issues. If we're going to make this a default, I think it's a good idea to minimize the possibility of things going wrong for now.

jose-elias-alvarez commented on lua/null-ls/builtins/code_actions/eslint_d.lua:
```diff
@@ -160,6 +162,10 @@ return h.make_builtin({
             return handler(params)
         end,
+        dynamic_command = cmd_resolver.from_node_modules,
+        cwd = function(params)
```
Setting an explicit `cwd` prevents these commands from taking advantage of the resolved `cwd` from the command resolver. That's both slower (since it'll search the filesystem each time) and less accurate (since it won't know about where the command was actually found).

In the future, I'm sure we can improve this mechanism to start searching in likelier locations, but for now I think it's best to leave it off.

kylo252 replied:
Sounds good! It's always tricky with caching, and I was actually worried if `cwd` would be able to get updated using my approach. I'll remove those for now.

kylo252 replied:
I removed the changes from `eslint_d` completely, since it will be handled in #521

jose-elias-alvarez commented on lua/null-ls/utils.lua:
```diff
@@ -62,8 +62,8 @@ M.has_version = function(ver)
 end
 
 M.is_executable = function(cmd)
-    local is_executable = vim.fn.executable(cmd) > 0
```
I'm curious – was this causing issues?

kylo252 replied:
It had wanted to re-order the check here: https://github.com/kylo252/null-ls.nvim/blob/6e9cbbebb6de8711c83c9a91603e0e716afe5bb3/lua/null-ls/helpers/command_resolver.lua#L39

Which meant that `cmd` could be `nil` which I'm not sure is actually safe to pass to `executable()`, but it's not really needed otherwise.

jose-elias-alvarez replied:
`vim.fn.executable(nil)` will throw an error, but perhaps we should handle that case specifically instead?

```lua
if not cmd then
    return false
end
```

jose-elias-alvarez reviewed (commented):
Apart from my comment I think this looks good. I'm curious to see the profiling results to see if it makes sense to set this as a default.

kylo252 commented:
> Apart from my comment I think this looks good. I'm curious to see the profiling results to see if it makes sense to set this as a default.

Any ideas for how to go about doing this profiling?

jose-elias-alvarez commented:
I'd use `vim.loop.now()` / `vim.loop.hrtime()` and capture 10-15 runs under the same environment, then average out the results (not sure if there is something more sophisticated).

kylo252 commented:
> I'd use `vim.loop.now()` / `vim.loop.hrtime()` and capture 10-15 runs under the same environment, then average out the results (not sure if there is something more sophisticated).

oh man, this took waaay longer than I expected, but the results were kinda wild as I was basically getting 300% improvement when running with a `dynamic_cmd`, but I can't say that I'm certain of my testing methodology since `vim.lsp.buf.formatting()` doesn't have a return value. 

So here are the test files that I used along with the results, please do try them and let me know what you think:
https://gist.github.com/kylo252/d2659cba93e4498c6d268ecce0848aa0

jose-elias-alvarez commented on lua/null-ls/helpers/command_resolver.lua:
```diff
@@ -0,0 +1,56 @@
+    -- a string means command was resolved on last run
+    -- false means the command already failed to resolve, so don't bother checking again
+    if resolved and (type(resolved.command) == "string" or resolved.command == false) then
+        log:debug(string.format("Using cached value for [%s] at %s", params.command, resolved.command))
```
Perhaps we could change this message when `resolved.command == false` to make it clearer that the command didn't resolve?

jose-elias-alvarez commented on lua/null-ls/helpers/command_resolver.lua:
```diff
@@ -0,0 +1,56 @@
+    if resolved_command then
+        local is_executable, err_msg = u.is_executable(resolved_command)
+        if not is_executable then
+            log:warn(err_msg)
```
We should throw an error here or handle this case to prevent caching a `resolved_command` that isn't executable.

jose-elias-alvarez commented:
The performance results are *really* strange – the difference isn't as dramatic on my setup, but both your script and manual tests confirm that `dynamic_command` is (after the first invocation) faster. I was able to replicate this with `stylua`, too, so doesn't seem to be just a `prettier` quirk. 

I won't pretend to understand how this can happen, but I guess we don't have to worry about performance concerns. I left a few additional comments, but other than what I mentioned I think this should be okay to merge.

kylo252 commented:
- I changed so that we're only assigning `resolved.command` to a valid value during the look up, this simplifies things overall
- I've added a separate log entry to indicate that a command wasn't resolved and won't be looked up again

Hopefully the logic is clear enough. Let me what you think.

jose-elias-alvarez commented on lua/null-ls/formatting.lua:
```diff
@@ -91,7 +91,9 @@ M.handler = function(method, original_params, handler)
             end,
             callback = function(edits, params)
-                M.apply_edits(edits, params)
+                if edits and #edits > 0 then
```
This is a sensible change but I'd prefer to leave it out for now, since #519 will change this entirely.

kylo252 replied:
Without this then we get  `received edits` even when the command was unable to unresolve

```logtalk
[DEBUG Thu 06 Jan 2022 05:19:16 PM CET] ...rt/null-ls.nvim/lua/null-ls/helpers/command_resolver.lua:18: attempting to find local executable node_modules/.bin/prettier
[WARN  Thu 06 Jan 2022 05:19:16 PM CET] ...rt/null-ls.nvim/lua/null-ls/helpers/command_resolver.lua:40: Unable to resolve command [node_modules/.bin/prettier], skipping further lookups
[DEBUG Thu 06 Jan 2022 05:19:16 PM CET] ...t/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:266: Using dynamic command for [prettier], got: nil
[DEBUG Thu 06 Jan 2022 05:19:16 PM CET] ...t/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:273: unable to resolve command [prettier]
[DEBUG Thu 06 Jan 2022 05:19:16 PM CET] ...ack/packer/start/null-ls.nvim/lua/null-ls/formatting.lua:55: received edits from generators
[DEBUG Thu 06 Jan 2022 05:19:16 PM CET] ...ack/packer/start/null-ls.nvim/lua/null-ls/formatting.lua:74: successfully applied edits
```

But this made me realize that the fallback solution I had is not perfect, because the generator attempted to spawn it without testing if it's available, (which is probably not too bad anyway, since it seems to timeout instead), this also without validating the `#edits`

```logtalk
[DEBUG Thu 06 Jan 2022 05:17:56 PM CET] ...rt/null-ls.nvim/lua/null-ls/helpers/command_resolver.lua:18: attempting to find local executable node_modules/.bin/prettier
[WARN  Thu 06 Jan 2022 05:17:56 PM CET] ...rt/null-ls.nvim/lua/null-ls/helpers/command_resolver.lua:40: Unable to resolve command [node_modules/.bin/prettier], skipping further lookups
[DEBUG Thu 06 Jan 2022 05:17:56 PM CET] ...t/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:266: Using dynamic command for [prettier], got: prettier
[DEBUG Thu 06 Jan 2022 05:17:56 PM CET] ...t/null-ls.nvim/lua/null-ls/helpers/generator_factory.lua:320: spawning command [prettier] at /tmp/test/nvim with args { "--stdin-filepath", "/home/hatsu/.local/share/nvim/lsp_servers/bash/node_modules/bash-language-server/node_modules/cssstyle/node_modules/prettier-linter-helpers/index.js" }
[DEBUG Thu 06 Jan 2022 05:18:01 PM CET] ...site/pack/packer/start/null-ls.nvim/lua/null-ls/loop.lua:138: command [prettier] timed out after 5000 ms
[DEBUG Thu 06 Jan 2022 05:18:01 PM CET] ...ack/packer/start/null-ls.nvim/lua/null-ls/formatting.lua:55: received edits from generators
[DEBUG Thu 06 Jan 2022 05:18:01 PM CET] ...ack/packer/start/null-ls.nvim/lua/null-ls/formatting.lua:74: successfully applied edits
```

jose-elias-alvarez replied:
This behavior is solved in #519, so I'd be happier leaving this out to prevent another rebase.

jose-elias-alvarez commented on lua/null-ls/helpers/command_resolver.lua:
```diff
@@ -0,0 +1,53 @@
+    local resolved = s.get_resolved_command(params.bufnr, params.command)
+    -- a string means command was resolved on last run
+    -- false means the command already failed to resolve, so don't bother checking again
+    if resolved and resolved.command then
```
Are we still handling the case where `resolved_command` was explicitly set to `false`? As far as I can tell we are running the scan again on each invocation, but the point of `false` was to indicate that we know the last attempt failed and we don't have to do it again.

kylo252 replied:
oops, I completely missed that part since I was focused on the log message. maybe it should be moved into the cache handler instead? or is it fine like this?

jose-elias-alvarez replied:
I don't think the cache handler should care about what's actually in the cache, so I think it's fine like this.

kylo252 commented:
> I also think we should mention this in the documentation.

Not sure how much details you want to add. I would also like to really encourage looking into #552

jose-elias-alvarez commented on doc/HELPERS.md:
```diff
@@ -104,8 +104,8 @@ original command (if set) is available as `params.command`. The callback should
 
-`dynamic_command` runs every time its parent generator runs and can affect
-performance, so it's best to cache its output when possible.
+The results of `dynamic_command` are cached since it runs every time its parent
```
I don't think this is accurate – the resolvers in this PR wrap `dynamic_command` to cache results and return them immediately, but this doesn't happen by default.

jose-elias-alvarez commented on lua/null-ls/helpers/command_resolver.lua:
```diff
@@ -0,0 +1,56 @@
+    end)
+
+    if not resolved.command then
+        log:warn(fmt("Unable to resolve command [%s], skipping further lookups", executable_to_find))
```
I think we should use `debug` here, since this is a normal situation when using `from_node_modules` and `prefer_local`.

jose-elias-alvarez commented on lua/null-ls/helpers/command_resolver.lua:
```diff
@@ -0,0 +1,56 @@
+M.generic = function(params, prefix)
+    local resolved = s.get_resolved_command(params.bufnr, params.command)
+    -- a string means command was resolved on last run
+    -- false means the command already failed to resolve, so don't bother checking again
```
We can remove or update this comment now since the type check is gone.

jose-elias-alvarez reviewed (commented):
A few minor changes left to make, but other than that I think this is ready.

jose-elias-alvarez commented:
Thanks! This went back-and-forth a bit, so could we just squash the commits into logical units before merging?

jose-elias-alvarez commented:
Thank you for all the work on this!
